File: src/com/leozhi/topic137/Solution03.java
package com.leozhi.topic137;

/**
 * @author leozhi
 * 位运算
 * 通过
 * 0ms
 */
public class Solution03 {
    public int singleNumber(int[] nums) {
        int seenOnce = 0;
        int seenTwice = 0;
        for (int num : nums) {
            seenOnce = ~seenTwice & (seenOnce ^ num);
            seenTwice = ~seenOnce & (seenTwice ^ num);
        }
        return seenOnce;
    }
}

File: src/com/leozhi/topic46/Solution03.kt
package com.leozhi.topic46

/**
 * @author leozhi
 * 回溯
 * 通过
 * 228ms
 */
class Solution03 {
    val res = ArrayList<ArrayList<Int>>()
    fun permute(nums: IntArray): List<List<Int>> {
        val track = ArrayList<Int>()
        backtrace(track, nums)
        return res
    }

    private fun backtrace(track: ArrayList<Int>, nums: IntArray) {
        if (track.size == nums.size) {
            res.add(ArrayList(track))
            return
        }
        for (num in nums) {
            if (track.contains(num)) {
                continue
            }
            track.add(num)
            backtrace(track, nums)
            track.remove(num)
        }
    }
}